File: examples/service_client_examples/usage_example.py
"""
Service Client Usage Example

This module demonstrates how to use the standardized service clients.
"""

import asyncio
import logging
from datetime import datetime, timedelta

from examples.service_client_examples.client_factory import (
    initialize_clients,
    get_market_data_client
)
from common_lib.clients.exceptions import (
    ClientError,
    ClientConnectionError,
    ClientTimeoutError,
    ClientValidationError,
    ClientAuthenticationError
)

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger("client_example")


async def market_data_example():
    """Example of using the Market Data client."""
    logger.info("Running Market Data client example...")
    
    # Get the client
    client = get_market_data_client()
    
    # Example 1: Get OHLCV data
    try:
        result = await client.get_ohlcv_data(
            symbol="EUR/USD",
            timeframe="1h",
            start_time=datetime.now() - timedelta(days=7),
            end_time=datetime.now(),
            limit=100
        )
        
        logger.info(f"Got {len(result.get('data', []))} OHLCV data points")
    except ClientTimeoutError as e:
        logger.error(f"Request timed out: {str(e)}")
    except ClientConnectionError as e:
        logger.error(f"Connection error: {str(e)}")
    except ClientError as e:
        logger.error(f"Client error: {str(e)}")
    
    # Example 2: Get instrument information
    try:
        instrument = await client.get_instrument("EUR/USD")
        logger.info(f"Got instrument information: {instrument}")
    except ClientError as e:
        logger.error(f"Failed to get instrument information: {str(e)}")
    
    # Example 3: List instruments
    try:
        instruments = await client.list_instruments(category="forex")
        logger.info(f"Got {len(instruments.get('data', []))} forex instruments")
    except ClientError as e:
        logger.error(f"Failed to list instruments: {str(e)}")
    
    # Example 4: Get latest price
    try:
        price = await client.get_latest_price("EUR/USD")
        logger.info(f"Latest EUR/USD price: {price}")
    except ClientError as e:
        logger.error(f"Failed to get latest price: {str(e)}")
    
    # Example 5: Using correlation ID
    try:
        # Create a client with correlation ID
        correlation_id = "example-correlation-id"
        client_with_correlation_id = await client.with_correlation_id(correlation_id)
        
        # Make a request with the correlation ID
        result = await client_with_correlation_id.get_ohlcv_data(
            symbol="EUR/USD",
            timeframe="1h",
            limit=10
        )
        
        logger.info(f"Got {len(result.get('data', []))} OHLCV data points with correlation ID")
    except ClientError as e:
        logger.error(f"Failed to get OHLCV data with correlation ID: {str(e)}")


async def main():
    """Main function."""
    logger.info("Starting service client example...")
    
    # Initialize clients
    initialize_clients()
    
    # Run examples
    await market_data_example()
    
    logger.info("Service client example completed")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/service_client_examples/usageExample.ts
/**
 * Service Client Usage Example
 * 
 * This module demonstrates how to use the standardized service clients.
 */

import { 
  initializeClients, 
  getMarketDataClient 
} from './clientFactory';

import {
  ClientError,
  ClientConnectionError,
  ClientTimeoutError,
  ClientValidationError,
  ClientAuthenticationError
} from '../../common-js-lib/index';

/**
 * Example of using the Market Data client
 */
async function marketDataExample() {
  console.log('Running Market Data client example...');
  
  // Get the client
  const client = getMarketDataClient();
  
  // Example 1: Get OHLCV data
  try {
    const startTime = new Date();
    startTime.setDate(startTime.getDate() - 7);
    
    const result = await client.getOHLCVData(
      'EUR/USD',
      '1h',
      startTime,
      new Date(),
      100
    );
    
    console.log(`Got ${result.data.length} OHLCV data points`);
  } catch (error) {
    if (error instanceof ClientTimeoutError) {
      console.error(`Request timed out: ${error.message}`);
    } else if (error instanceof ClientConnectionError) {
      console.error(`Connection error: ${error.message}`);
    } else if (error instanceof ClientError) {
      console.error(`Client error: ${error.message}`);
    } else {
      console.error(`Unexpected error: ${error}`);
    }
  }
  
  // Example 2: Get instrument information
  try {
    const instrument = await client.getInstrument('EUR/USD');
    console.log(`Got instrument information: ${JSON.stringify(instrument)}`);
  } catch (error) {
    console.error(`Failed to get instrument information: ${error}`);
  }
  
  // Example 3: List instruments
  try {
    const instruments = await client.listInstruments('forex');
    console.log(`Got ${instruments.data.length} forex instruments`);
  } catch (error) {
    console.error(`Failed to list instruments: ${error}`);
  }
  
  // Example 4: Get latest price
  try {
    const price = await client.getLatestPrice('EUR/USD');
    console.log(`Latest EUR/USD price: ${JSON.stringify(price)}`);
  } catch (error) {
    console.error(`Failed to get latest price: ${error}`);
  }
  
  // Example 5: Using correlation ID
  try {
    // Create a client with correlation ID
    const correlationId = 'example-correlation-id';
    const clientWithCorrelationId = client.withCorrelationId(correlationId);
    
    // Make a request with the correlation ID
    const result = await clientWithCorrelationId.getOHLCVData(
      'EUR/USD',
      '1h',
      undefined,
      undefined,
      10
    );
    
    console.log(`Got ${result.data.length} OHLCV data points with correlation ID`);
  } catch (error) {
    console.error(`Failed to get OHLCV data with correlation ID: ${error}`);
  }
}

/**
 * Main function
 */
async function main() {
  console.log('Starting service client example...');
  
  // Initialize clients
  initializeClients();
  
  // Run examples
  await marketDataExample();
  
  console.log('Service client example completed');
}

// Run the example
main().catch(error => {
  console.error('Error running example:', error);
});